Automate creating a GH release

DennisDenuto opened the pull request:
Use [goreleaser](https://goreleaser.com/) to create a GH release

Use [github-release-notes](github.com/buchanae/github-release-notes) to generate release notes

TODO:

- [x] add goreleaser github action (trigger a release when a tag is pushed)
- [x] Upload binaries (instead of uploading the tar.gz files)
- [x] Checksum file should include checksums for binaries (not the tar.gz files)
- [x] ensure that binary files are 'reproducible builds' (goreleaser builds the same sha as ./hack/build-binaries.sh)
- [x] Use ld flags to update the version variable (from a tag) when publishing a release
~~- [x] make a carvel member generate a one-time token to provide to the goreleaser github action~~ <- No need. read: https://docs.github.com/en/actions/reference/authentication-in-a-workflow#about-the-github_token-secret
- [x] Release notes should contain checksum (not just as an attached file) <- leave this as a manual step for now
- [x] add a post release step in the GH action, that verifies the SHA of the binaries uploaded via gorelease match up with the SHAs built by a native go binary
- [x] change `hack/build-binary.sh` to generate the version based on the tag the commit is on. since version is now a build flag.
- [x] check if we can restrict network of the github action (only allow requests to github.com)??
- [x] Change github owner (in the .goreleaser.yml file) from DennisDenuto -> vmware-tanzu
~~- [ ] hack/build scripts can use goreleaser to generate binary files~~
- [ ] automate homebrew update https://goreleaser.com/customization/homebrew/
- [ ] automate creating a snapshot of OSM latest
~~- [ ] copy the gorelease GH action into our repo. and in our gh.yml point to the digest of the relocated image~~
- [ ] Update release documentation
  - [ ] review draft release notes and add the checksum to the release notes (goreleaser attaches the file but doesn't add it to the release notes section)

DennisDenuto committed:
Automate creating a GH release

- Add goreleaser configuration file
- Add github action that will run gorelease when a tag is pushed

Authored-by: Dennis Leon <leonde@vmware.com>

danielhelfand reviewed (commented):
The goreleaser.yml looks great. 

Was thinking a bit more on triggering the release. Was wondering if changing the version change commit step to something like `go build -ldflags "-X github.com/vmware-tanzu/carvel-imgpkg/pkg/imgpkg/cmd/version.Version=$RELEASE_VERSION"` and invoking whatever script is used [manually via a GitHub Action](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow). The user could just specify the version to the action and run it.

DennisDenuto commented:
@danielhelfand I like this idea. I'll add it to the goreleaser.yml file. I'll be working on this prob this Friday.

DennisDenuto commented:
Adding some concerns around security raised by @jtigger 

1. tons of stuff around GH (including getting into our GitHub Action container)
1. goreleaser itself injecting malicious code
1. modifying the goreleaser/goreleaser image, injecting malicious code


A few ideas towards making this more 'secure':
1. Use the native go binary to build imgpkg and only use goreleaser to update the gh release notes / attach the files built by the native go binary
2. Add a post release step in the GH action, that verifies the SHA of the binaries uploaded via gorelease match up with the SHAs built by a native go binary 
3. relocate/fork the goreleaser github action, into our dockerhub repo to ensure we have more control over it. (At time of writing, to ease with development I am using the public goreleaser github action)

DennisDenuto commented:
Spoke to @jtigger 

We decided on:
- [x] make a carvel member generate a one-time token to provide to the goreleaser github action
- [x] add a post release step in the GH action, that verifies the SHA of the binaries uploaded via gorelease match up with the SHAs built by a native go binary
- [x] check if we can restrict network of the github action (only allow requests to github.com)??


Couldn't find a way to restrict network access on GH actions. However further reduced the scope of our token to 'write' on content. Allowing it enough permission to create a release.

DennisDenuto committed:
Set version of imgpkg based on the tag

DennisDenuto committed:
Add verification step to ensure the integrity of what was uplaoded to github

- check if goreleaser binaries match what is generated by hack/build scripts

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,98 @@
+            const https = require('https');
+
+            checksums = {}
+            for (const r of releases["data"]) {
```
this seems a bit backwards to me -- instead of being an authority for checksums for built binaries, we are asking the system to give us binaries back to generate the checksums. we are trusting that upload succeeded, nothing was tampered in any way, and we have successfully downloaded everything.

mya be another angle to think about is how does signing play into this (since we would want to sign a set of checksums)? doesnt goreleaser generate checksums.txt?

DennisDenuto replied:
hmmm, this part of the release GH action is to verify that the binaries uploaded by goreleaser match the same sha as binaries generated via the `hack/build-binaries.sh`

The concern is that `goreleaser` *could* have been compromised, and thus, uploaded a compromised binary. 

So, compare the sha's of the uploaded binary built with `goreleaser`, with the sha's generated using native `go build` via `hack/build-binaries.sh`

> upload succeeded , nothing was tampered

Perhaps I should just assume nothing was tampered. Although leaving this 'verification' step is hopefully one less attack vector.


FWIW, This decision was largely driven by the above comment: https://github.com/vmware-tanzu/carvel-imgpkg/pull/117#issuecomment-819869853

cppforlife replied:
> So, compare the sha's of the uploaded binary built with goreleaser, with the sha's generated using native go build via hack/build-binaries.sh

oh very interesting. i think i missed that part. are we able to reliably guarantee that our script is equivalent to whatever goreleaser is producing?

DennisDenuto replied:
> our script is equivalent to whatever goreleaser is producing

Yes, locally and when run as a GH action it produces the same `sha` as our internal `hack/build.sh` script.

also, from the debug logs goreleaser provides, it runs the equivalent `go build` command...

However, `goreleaser` is a binary. And while today it generates the same `sha` it might not tomorrow (assuming it is compromised / malicious)


> i think i missed that part

This script `get-checksums-from-draft-release` downloads the binaries attached to the github draft release and calculates its `sha`. The next step compares the `sha` generated by `hack/build` with the `sha` calculated in this step of the GH action.

By the time this step is running, the token given to `goreleaser` (previous step) should be revocated. i.e. `goreleaser` can't send that token to another server to upload a different binary to our GH release in some future date (during and after this verification step runs)

DennisDenuto replied:
Had a call with dk. Marking this conversation as resolved.

cppforlife commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,98 @@
+          go-version: 1.16.0
+      -
+        name: Run GoReleaser
+        uses: goreleaser/goreleaser-action@v2
```
should we be locking down to a particular digest? (this would help us to prevent a threat where goreleaser repo is compramised)

DennisDenuto replied:
this is a good idea. The plan was actually to fork and host our own copy of `goreleaser-action`

mainly driven by: https://github.com/vmware-tanzu/carvel-imgpkg/pull/117#issuecomment-819869853

Although I think for now, a sha is more then sufficient.

DennisDenuto replied:
Done

cppforlife commented on .goreleaser.yml:
```diff
@@ -0,0 +1,84 @@
+
+    ldflags:
+      - -buildid=
+      - -X github.com/k14s/imgpkg/pkg/imgpkg/cmd.Version={{ .Version }}
```
what are thoughts behind injecting version here vs in code?

DennisDenuto replied:
It results in 1 less manual step done during a release.

Instead of making a code change and then pushing a tag. It changes to: 'push a tag'

If someone wanted to checkout the code and build our project, they can rebuild the *same* binary uploaded via `goreleaser` or via the `hack/build-binary.sh`.

DennisDenuto committed:
change hack/build.sh to retrieve the last released tag as the version to use when building locally

This mirrors what was done previously, when a code change was made to
versions.go. (Using the last released tag version)

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
restrict permissions to only 'write' access for releases

If you specify the access for any of these scopes, all of those that are
not specified are set to none. For more information:
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#permissions

For more information on what 'contents' permission entails:
https://docs.github.com/en/rest/reference/permissions-required-for-github-apps#permission-on-contents

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Pin the version of goreleaser action (v2) to a commit sha instead

- sha is used from: https://github.com/goreleaser/goreleaser-action/releases/tag/v2

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Update github owner to vmware-tanzu

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto commented:
Will move the below remaining TODO items in a separate (future PR)

```
- automate homebrew update https://goreleaser.com/customization/homebrew/
- automate creating a snapshot of OSM latest
- Update release documentation
```

cppforlife commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,100 @@
+name: goreleaser
+
+on:
+  pull_request:
```
do we need this for PRs?

DennisDenuto replied:
nope. removed.

DennisDenuto committed:
Set the default version of imgpkg to 'dev'

- the version is determined at build time and based off the tag the
release is run against

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on .gitignore:
```diff
@@ -5,3 +5,4 @@
 /tmp
 .DS_Store
 /test/e2e/assets/simple-app/empty-dir
+dist/
```
should we move this into tmp/dist?

DennisDenuto replied:
hmm... I don't see a config for goreleaser to make this change. 
but also, goreleaser will practically only ever be run by an ephemeral env (github action). So the `dist` folder should rarely (if ever) make it on a dev workstation.

cppforlife commented on hack/build.sh:
```diff
@@ -2,16 +2,18 @@
 
 set -e -x -u
 
+VERSION="$(git describe --tags | grep -Eo 'v[0-9]+\.[0-9]+\.[0-9]+' || echo 'develop')"
```
any reason to bother to include this in hack/build.sh? it's only used for development.

DennisDenuto replied:
nope. changed.

cppforlife commented on pkg/imgpkg/cmd/version.go:
```diff
@@ -10,9 +10,7 @@ import (
-const (
-	Version = "0.8.0"
-)
+var Version = ""
```
lets set this to `develop`

DennisDenuto replied:
done

DennisDenuto committed:
Minor fix

- Don't run releaser GH action on pull requests
- change hack/build.sh to not provide version via ldflag.
-- it is only run by dev to run tests. Leaving the version to 'develop'
and not the latest tag name is fine and results in less bash code to
maintain

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on hack/build-binaries.sh:
```diff
@@ -2,14 +2,20 @@
+LATEST_GIT_TAG=$(git describe --tags | grep -Eo 'v[0-9]+\.[0-9]+\.[0-9]+')
+VERSION="${1:-$LATEST_GIT_TAG}"
+
+go fmt ./cmd/... ./pkg/... ./test/...
+go mod vendor
+go mod tidy
```
just noticed this, do we need to do this here?
